text/3349-cfg-os-version.md

```diff
@@ -0,0 +1,142 @@
+
+If a library does not explicitly set its `min_os_target_windows` value, it will automatically be set to the largest `min_windows_build_version` of all of its transitive dependencies.
+
+The `os_version_min` predicate evaluates to `true` when the provided string is greater than or equal to the target OS version and the marked code will be compiled.
+
+For targets where `os_version_min(“windows”, …)` does not make sense (i.e., non-Windows targets), the `cfg` predicate will return `false` and emit a warning saying that the particular `cfg` predicate is not supported on that target. Therefore, it's important to pair `os_version_min(“windows”, …)` with a `cfg(windows)` using the existing mechanisms for combining `cfg` predicates.
+
+The above example works exactly the same way with the other platform API `cfg` predicates just with different values and different target support.
+
+These predicates do not assume any semantic versioning information. The specified predicates are simply listed in order. The only semantics that are assumed is that code compiled with the `cfg` predicates works for all versions greater than or equal to that version.
+
+**Note:** Here it would be important to link to documentation showing the `cfg` predicates and the different version strings that are supported.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The various target API version `cfg` predicates allow users to conditionally compile code based on the API version supported by the target platform.  Each platform is responsible for defining a default key, a set of keys it supports, and functions that are able to compare the version strings they use.  A set of comparison functions can be provided by `rustc` for common formats such as 2- and 3-part semantic versioning.  When a platform detects a key it doesn’t support it will return `false` and emit a warning.
+
+When a target is being built the actual API versions will be set via the following methods, in decreasing order of precedence:
+* Command line
+* Cargo.toml target sections
+* Target platform defaults
+
+## Versioning Schema
+
+Version strings can take on nearly any form and while there are some standard formats, such as semantic versioning or release dates, projects can change schemas or provide aliases for some or all of their releases.  Because of this diversity in version strings each platform will be responsible for defining a type implementing `FromStr`, `Display`, and `Ord` for each key they support (or using one of the pre-defined types).
+
+## Storing Information in Metadata
+
+The values for the various `cfg` requirements must be stored in a crate’s metadata if the value is set.  If no value is set the platform’s default API versions are used.  If no value is set it will default to the maximum of all the crate’s transitive dependencies’ requirements.
```

Comment:
If I understand this clause correctly then this feature have little resemblance to regular `cfg`.
In particular it cannot be used during macro expansion / "preprocessing" and requires something like monomorphisation to implement.
Regular `cfg` cannot do any of that.

We have another feature with similar semantics - link time `cfg` (`#[link(..., cfg(predicate))]`) for which the predicate is evaluated and used only when the final binary (e.g. executable or a shared library) is linked.

A target version cfg evaluated only when producing a "final binary" would certainly be a much more useful feature, in particular it would not require rebuilding standard library (and would be more similar to version headers in C in general), but that's a separate feature that's only similar to regular expansion time `cfg` because it can evaluate predicates.

Reply:
I'm afraid I don't see why this would require monomorphisation.  It seems that if the requirements are recorded in the crate metadata rustc can verify that they are met when producing new binaries.  If there is a mismatch in requirements an error is produced, otherwise we link normally.

I do agree that being able to evaluate the condition at link time would be ideal.  Does the compiler currently support anything like that?